src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,845 @@
+            new_attention_prob = torch.zeros(2, attention_prob_0.shape[1], max_dim2, max_dim3).to(
+                attention_prob_0.device
+            )
+            new_attention_prob[0, :, : attention_prob_0.shape[2], : attention_prob_0.shape[3]] = attention_prob_0
+            new_attention_prob[1, :, : attention_prob_1.shape[2], : attention_prob_1.shape[3]] = attention_prob_1
+            new_attention_probs = new_attention_probs + (new_attention_prob,)
+        else:
+            new_attention_probs = new_attention_probs + (torch.cat([attention_prob_0, attention_prob_1]),)
+    return new_attention_probs
+
+
+def stack_attention_probs_list(attention_probs: List[torch.Tensor]) -> torch.Tensor:
+    all_attention_probs_shape_the_same = True
+    first_attention_prob_shape = attention_probs[0].shape
+    for attention_prob in attention_probs[1:]:
+        if attention_prob.shape != first_attention_prob_shape:
+            all_attention_probs_shape_the_same = False
+    if all_attention_probs_shape_the_same:
+        stacked_attention_probs = torch.stack(attention_probs, dim=0)
+    else:
+        max_dim2 = max([attention_prob.shape[2] for attention_prob in attention_probs])
+        max_dim3 = max([attention_prob.shape[3] for attention_prob in attention_probs])
+        stacked_attention_probs = torch.zeros(
+            len(attention_probs), 2, attention_probs[0].shape[1], max_dim2, max_dim3
+        ).to(attention_probs[0].device)
+        for i, attention_prob in enumerate(attention_probs):
+            stacked_attention_probs[i, :, :, : attention_prob.shape[2], : attention_prob.shape[3]] = attention_prob
+    return stacked_attention_probs
+
+
+def batch_attention_probs_list(
+    attention_probs: Union[List[torch.Tensor], List[Tuple[torch.Tensor]]]
+) -> Union[List[torch.Tensor], List[Tuple[torch.Tensor]]]:
+    """
+    Given a list of attention probabilities, batch them together.
+    We assume that the attention probabilities are of shape (batch_size, num_heads, num_keypoints_a, num_keypoints_b).
+    The list can be of two forms:
+    - List of attention probabilities: [attention_probs_0, attention_probs_1, ...]
+    - List of tuples of attention probabilities: [(attention_probs_0_0, attention_probs_1_0, ...), (attention_probs_0_1, attention_probs_1_1, ...), ...]
+    If the list is of the first form, we stack the attention probabilities along the batch dimension:
+    [attention_probs_0, attention_probs_1, ...] -> torch.stack([attention_probs_0, attention_probs_1, ...], dim=0)
+    If the list is of the second form, we stack the attention probabilities along the batch dimension for each tuple:
+    [(attention_probs_0_0, attention_probs_1_0, ...), (attention_probs_0_1, attention_probs_1_1, ...), ...] ->
+    [torch.stack([attention_probs_0_0, attention_probs_1_0, ...], dim=0), torch.stack([attention_probs_0_1, attention_probs_1_1, ...], dim=0), ...]
+    """
+    if len(attention_probs) > 0:
```

Comment:
Do we need to handle all these cases?

Reply:
While I was writing this function I encountered such cases. Fast forward to this commit, turns out this function is either called with a `list = [None * batch_size]` which returns None (which obviously depends on the `output_attention` argument being equal to False), or `list=[tuple(tensors) * batch_size]`, and actually never fall into the case L105. I got rid of some of these lines in [this commit](https://github.com/huggingface/transformers/pull/29886/commits/25fd37d15e16a2600a7c0eaa052890d81177580a)